clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,119 @@
+// RUN: %clang_cc1 %s -fsyntax-only -ast-dump -fsycl-is-device -triple spir64 -verify
+
+// expected-error 4 @+2{{kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+template <typename KernelName, typename KernelType>
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) {
+  kernelFunc();
+}
+template <template <typename> typename Outer, typename Inner>
+struct S {
+  void operator()() const;
+};
+
+template <typename Ty>
+struct T {};
+
+template <typename Func>
+void kernel3func(const Func &F3) {
+  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  constexpr const char *F3_output = __builtin_unique_stable_name(F3);
+  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+  kernel_single_task<class kernel3>(F3);
+}
+
+template <typename Func>
+void kernel4func(const Func &F4) {
+  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  constexpr const char *F4_output = __builtin_unique_stable_name(F4);
+  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+  kernel_single_task<class kernel4>([]() {});
+}
+
+template <typename Ty>
+auto func() -> decltype(__builtin_unique_stable_name(Ty::str));
+
+struct Derp {
+  static constexpr const char str[] = "derp derp derp";
+};
+
+template <typename T>
+static constexpr const char *output1 = __builtin_unique_stable_name(T);
+
+#define MACRO12()                \
+  auto l12 = []() { return 1; }; \
+  constexpr const char *l1_output = __builtin_unique_stable_name(l12);
+
+int main() {
+  // kernel0
+  kernel_single_task<class kernel0>( // func<Derp>());
+      []() {
+        func<Derp>();
+      });
+
+  // kernel1 Evaluate the builtin in a constant context, then call the kernel with that same type
+  auto l1 = []() {};
+  constexpr const char *l1_output = __builtin_unique_stable_name(l1);
+  kernel_single_task<class kernel1>(
+      [=]() { l1(); });
+
+  // kernel2 Call to the builtin on a lambda, then using the lambda for a kernel name
+  auto l2 = []() { return 1; };
+  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  constexpr const char *l2_output = __builtin_unique_stable_name(l2);
+  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+  kernel_single_task<class kernel2>(l2);
+
+  // kernel3
+  // The current function is named with a lambda (that is, takes a lambda as a template parameter).  Call the builtin (perhaps on the current function?), then pass it to a kernel.
+  kernel3func([]() {});
```

Comment:
The comments don't seem to match the code here -- the current function is `main`, which definitely does not take a lambda.

Reply:
I see. I took it to mean "a function"

Reply:
Function is somewhat correct -- it means:
```
template <typename Callable>
void func(const Callable &C); // Function has a template parameter

void foo() {
  func([](){}); // Called with a lambda
}

void bar() {
  func(foo); // Called without a lambda
}
```